plugins/samples/enable_recaptcha/plugin.rs

```diff
@@ -0,0 +1,221 @@
+    // HtmlRewriter. Member of MyHttpContext a.k.a "StreamContext" so that the
+    // rewriter persists across multiple body callbacks.The rewriter parses
+    // and modifies chunks of Html. Behavior for Html modifications are defined
+    // via content_handlers. The rewriter only writes completed sections of Html to
+    // the output sink(e.g., when the rewriter parses "<di", nothing will be
+    // sent to the output sink. If the next chunk was "v> <h1>foo</h", the
+    // rewriter would then write "<div><h1>foo" to the output sink) unless
+    // HtmlRewriter::end() is called. HtmlRewriter::end() results in all
+    // uncompleted Html being buffered in rewriter to be sent to output sink as
+    // if it were plain text.
+    rewriter: Option<HtmlRewriter<'a, MyOutputSink>>,
+    // True when plugin has added script to <head>.
+    completed_script_injection: Rc<RefCell<bool>>,
+    // reCaptcha site key
+    recaptcha_key: Rc<RefCell<String>>,
+}
+
+impl<'a> MyHttpContext<'a> {
+    pub fn new(key: Rc<RefCell<String>>) -> MyHttpContext<'a> {
+        let output = Rc::new(RefCell::new(Vec::new()));
+        let completed_script_injection = Rc::new(RefCell::new(false));
+        let recaptcha_key = key;
+        MyHttpContext {
+            output: output.clone(),
+            completed_script_injection: completed_script_injection.clone(),
+            recaptcha_key: recaptcha_key.clone(),
+            rewriter: Some(HtmlRewriter::new(
+                Settings {
+                    element_content_handlers: vec![element!("head", move |el| {
+                        el.prepend(
+                            format!("\n<script src=\"https://www.google.com/recaptcha/enterprise.js?render={}\"></script>\n", *recaptcha_key.borrow_mut()).as_str(),
+                            ContentType::Html,
+                            );
+                        *completed_script_injection.borrow_mut() = true;
+                        Ok(())
+                    })],
+                    ..Settings::new()
+                },
+                MyOutputSink {
+                    output_sink: output,
+                },
+            )),
+        }
+    }
+
+    fn parse_chunk(&mut self, body_bytes: Bytes) -> Result<(), Box<dyn Error>> {
+        // Parse/rewrite current chunk
+        self.rewriter
+            .as_mut()
+            .ok_or("Expected valid rewriter. Got None")?
+            .write(&body_bytes)?;
+        return Ok(());
+    }
+
+    fn end_rewriter(&mut self) -> Result<(), Box<dyn Error>> {
+        // Stop the rewriter after completing desired domain rewrite to
+        // dump any unparsable inputs to output.
+        self.rewriter
+            .take()
+            .expect("Expected valid rewriter. Got None")
+            .end()?;
+        return Ok(());
+    }
+}
+
+impl<'a> Context for MyHttpContext<'a> {}
+
+impl<'a> HttpContext for MyHttpContext<'a> {
+    // If request header callback is invoked, body will not be compressed.
```

Comment:
What causes this to be invoked or not invoked? It doesn't appear to be configurable through the plugin configuration.

Reply:
Callbacks are invoked based on Envoy configuration. This repo does not replicate this functionality. This was only included in case users wanted the ability to disable compression.

Reply:
Removing, per discussion